File: app/moderation_provider.py
from __future__ import annotations

import json
import os
import time
import uuid
from dataclasses import dataclass
from typing import Any

import httpx


def _extract_first_json_object(text: str) -> dict[str, Any] | None:
    if not isinstance(text, str) or not text.strip():
        return None

    trimmed = text.strip()
    try:
        parsed = json.loads(trimmed)
        return parsed if isinstance(parsed, dict) else None
    except json.JSONDecodeError:
        pass

    start = trimmed.find("{")
    end = trimmed.rfind("}")
    if start < 0 or end <= start:
        return None

    candidate = trimmed[start : end + 1]
    try:
        parsed = json.loads(candidate)
        return parsed if isinstance(parsed, dict) else None
    except json.JSONDecodeError:
        return None


def _normalize_verdict(value: Any) -> str:
    normalized = str(value or "").lower()
    if normalized in {"allow", "flag", "block"}:
        return normalized
    return "flag"


@dataclass(slots=True)
class ModerationProvider:
    name: str

    async def moderate(self, payload: dict[str, Any]) -> dict[str, Any]:
        raise NotImplementedError


class MockModerationProvider(ModerationProvider):
    def __init__(self) -> None:
        super().__init__(name="mock")

    async def moderate(self, payload: dict[str, Any]) -> dict[str, Any]:
        return {
            "messageId": payload.get("messageId") or str(uuid.uuid4()),
            "verdict": "allow",
            "confidence": 1.0,
            "category": "safe",
            "reason": "placeholder-verdict",
            "latencyMs": 1,
        }


class OllamaModerationProvider(ModerationProvider):
    def __init__(self, base_url: str, model: str, timeout_ms: int) -> None:
        super().__init__(name="ollama")
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.timeout_ms = timeout_ms

    async def moderate(self, payload: dict[str, Any]) -> dict[str, Any]:
        started = time.perf_counter()
        prompt_payload = {
            "platform": payload.get("platform", "unknown"),
            "username": payload.get("username", "unknown"),
            "text": payload.get("text", ""),
        }
        system_prompt = (
            "You are a strict chat moderation classifier. "
            "Return only valid JSON with keys: verdict, confidence, category, reason."
        )

        async with httpx.AsyncClient(timeout=self.timeout_ms / 1000) as client:
            response = await client.post(
                f"{self.base_url}/api/generate",
                json={
                    "model": self.model,
                    "stream": False,
                    "format": "json",
                    "prompt": f"{system_prompt}\n\nInput:\n{json.dumps(prompt_payload, ensure_ascii=False)}",
                },
            )
            response.raise_for_status()
            body = response.json()

        parsed = _extract_first_json_object(body.get("response", ""))
        if parsed is None:
            raise ValueError("Ollama response did not contain a valid JSON object.")

        latency_ms = int((time.perf_counter() - started) * 1000)
        confidence_raw = parsed.get("confidence")
        confidence = float(confidence_raw) if isinstance(confidence_raw, (int, float)) else 0.5

        return {
            "messageId": payload.get("messageId") or str(uuid.uuid4()),
            "verdict": _normalize_verdict(parsed.get("verdict")),
            "confidence": confidence,
            "category": str(parsed.get("category") or "unknown"),
            "reason": str(parsed.get("reason") or "model-response"),
            "latencyMs": latency_ms,
        }


def create_moderation_provider() -> ModerationProvider:
    provider_name = os.getenv("LLM_PROVIDER", "mock").lower()
    if provider_name == "ollama":
        return OllamaModerationProvider(
            base_url=os.getenv("OLLAMA_BASE_URL", "http://127.0.0.1:11434"),
            model=os.getenv("OLLAMA_MODEL", "qwen2.5:7b"),
            timeout_ms=int(os.getenv("LLM_TIMEOUT_MS", "6000")),
        )

    return MockModerationProvider()

File: src/moderationProvider.js
import crypto from "node:crypto";

function createMockProvider() {
  return {
    name: "mock",
    async moderate(payload) {
      return {
        messageId: payload?.messageId || crypto.randomUUID(),
        verdict: "allow",
        confidence: 1,
        category: "safe",
        reason: "placeholder-verdict",
        latencyMs: 1,
      };
    },
  };
}

function extractFirstJsonObject(input) {
  if (!input || typeof input !== "string") return null;

  const trimmed = input.trim();
  try {
    return JSON.parse(trimmed);
  } catch {}

  const start = trimmed.indexOf("{");
  const end = trimmed.lastIndexOf("}");
  if (start < 0 || end <= start) return null;

  const candidate = trimmed.slice(start, end + 1);
  try {
    return JSON.parse(candidate);
  } catch {
    return null;
  }
}

function normalizeVerdict(value) {
  const normalized = String(value || "").toLowerCase();
  if (normalized === "allow" || normalized === "flag" || normalized === "block") return normalized;
  return "flag";
}

function createOllamaProvider({
  baseUrl,
  model,
  timeoutMs = 6000,
  fetchImpl = fetch,
  logger = console,
}) {
  const endpoint = `${baseUrl.replace(/\/$/, "")}/api/generate`;

  return {
    name: "ollama",
    async moderate(payload) {
      const start = Date.now();
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeoutMs);

      const systemPrompt =
        "You are a strict chat moderation classifier. Return only valid JSON with keys: verdict, confidence, category, reason.";
      const userPrompt = JSON.stringify(
        {
          platform: payload?.platform || "unknown",
          username: payload?.username || "unknown",
          text: payload?.text || "",
        },
        null,
        2,
      );

      try {
        const response = await fetchImpl(endpoint, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          signal: controller.signal,
          body: JSON.stringify({
            model,
            stream: false,
            format: "json",
            prompt: `${systemPrompt}\n\nInput:\n${userPrompt}`,
          }),
        });

        if (!response.ok) {
          throw new Error(`Ollama request failed with ${response.status}`);
        }

        const body = await response.json();
        const parsed = extractFirstJsonObject(body?.response);
        if (!parsed) {
          throw new Error("Ollama response did not contain valid JSON result");
        }

        return {
          messageId: payload?.messageId || crypto.randomUUID(),
          verdict: normalizeVerdict(parsed.verdict),
          confidence: Number.isFinite(parsed.confidence) ? parsed.confidence : 0.5,
          category: String(parsed.category || "unknown"),
          reason: String(parsed.reason || "model-response"),
          latencyMs: Date.now() - start,
        };
      } finally {
        clearTimeout(timer);
      }
    },
  };
}

export function createModerationProvider({
  provider = process.env.LLM_PROVIDER || "mock",
  logger = console,
  fetchImpl = fetch,
  baseUrl = process.env.OLLAMA_BASE_URL || "http://127.0.0.1:11434",
  model = process.env.OLLAMA_MODEL || "qwen2.5:7b",
  timeoutMs = Number.parseInt(process.env.LLM_TIMEOUT_MS || "6000", 10),
} = {}) {
  if (provider.toLowerCase() === "ollama") {
    logger.info(
      `Using ollama moderation provider (${model}) at ${baseUrl}, timeout ${timeoutMs}ms`,
    );
    return createOllamaProvider({ baseUrl, model, timeoutMs, fetchImpl, logger });
  }

  logger.info("Using mock moderation provider.");
  return createMockProvider();
}
